File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        q = deque([root])

        while q:
            qLen = len(q)
            rightSide = None
            for i in range(qLen):
                curr = q.popleft()
                if curr:
                    rightSide = curr
                    q.append(curr.left)
                    q.append(curr.right)
            if rightSide:
                res.append(rightSide.val)
        
        return res

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res = 0;
    int goodNodes(TreeNode* root) {
        dfs(root, root->val);
        return res;
    }

    void dfs(TreeNode* node, int path_max) {
        if (!node) return;

        if (node->val >= path_max) res++;
        dfs(node->left, max(path_max, node->val));
        dfs(node->right, max(path_max, node->val));
    }
};
